### ISSUE TITLE
XFCE tumbler does not work

### ISSUE THREAD

**Issue by bell07** [GerbilSoft/rom-properties#245]
The thumbnails are not shown in XFCE thunar, because the cached file is empty.
Example: 
```
$ /usr/libexec/rp-download -v gba/title/E/ASOE78.png
URL: https://rpdb.gerbilsoft.com/gba/title/E/ASOE78.png
Cache Filename: /home/bell07/.cache/rom-properties/gba/title/E/ASOE78.png
/usr/libexec/rp-download: Error downloading file: File exists
```
An empty file /home/bell07/.cache/rom-properties/gba/title/E/ASOE78.png was created trough this call.

My thumbler version is gentoo standard xfce-extra/tumbler-0.2.8 with USE="curl ffmpeg jpeg pdf -gstreamer -odf -raw"
Site note: Without curl support (gentoo default) the error message was "Invalid system call"..

**Comment by Amnesia1000** [comment]
I had the same problem, it is resolved if you install the latest update from the Git, the issue is #239 (I recommend you read it and be aware). If you do, it will correctly download the cover.

The rom-properties cache is now stored in "/usr/share/rom-properties/cache/" and has to be moved manually. But currently it is not working either.

It is already working (I suppose) to make it work, it only remains to wait.

If it doesn't work when updating from the git, I recommend that you comment on it in the issue mentioned above.

**Comment by GerbilSoft** [comment]
I've noticed that on my system (also Gentoo), Thunar doesn't use tumbler in some cases. In particular, it seems that it prefers GNOME thumbnailers.

Regardless: Can you rebuild with `-DENABLE_SECCOMP_DEBUG=1` and post any syscall trap messages printed by rp-download? These may have been fixed in git master already.

**Comment by bell07** [comment]
Hm, /usr/share/rom-properties/cache/ does not exists. Usually /usr/share is not writeable by non-root, so it is the wrong place for caches.
I try to adjust my ebuild for debug tomorrow.
For tumbs my gentoo uses [xfce-extra/tumbler-0.2.8](https://docs.xfce.org/xfce/thunar/tumbler)
See now some files does have tumbs (NDS, 3DS) ...but other (GB, GBA) does not :-/

**Comment by Amnesia1000** [comment]
I know, if you read the issue that I mentioned above you will understand the reason for everything. It is not a homemade fix, it is how you can solve the problem you have by the developer himself (if it is the same, since in the issue already mentioned the problem is by gnome, but it could be the same)

**Comment by Amnesia1000** [comment]
> Hm, /usr/share/rom-properties/cache/ does not exists. Usually /usr/share is not writeable by non-root, so it is the wrong place for caches.
> I try to adjust my ebuild for debug tomorrow.
> For tumbs my gentoo uses [xfce-extra/tumbler-0.2.8](https://docs.xfce.org/xfce/thunar/tumbler)
> See now some files does have tumbs (NDS, 3DS) ...but other (GB, GBA) does not :-/

The 3DS and DS thumbnails are taken internally from the game, but the GBA and GB thumbnails are downloaded from the internet, so you can see the 3DS and DS thumbnails and not the GBA and GB thumbnails.

Also here you will find information that in certain cases (it could be yours) the /usr/share/rom-properties/cache/ folder could be used. [https://github.com/GerbilSoft/rom-properties/blob/master/NEWS.md#v17-released-2020](https://github.com/GerbilSoft/rom-properties/blob/master/NEWS.md#v17-released-2020) in the next changes for version 1.7

If you can fix the problem, could you tell me (if you have activated using covers on consoles like Wii, 3DS, DS or Gamecube) how the cover folders were organized, because the problem of downloading covers was solved, but I still didn't It shows them, and since the copy of files to the /usr/share/rom-properties/cache/ folder must be manual, I don't know if they are correctly placed.
The folder path has remained like this:
/usr/share/rom-properties/cache/3ds/coverM/US/XXXX.jpg
/usr/share/rom-properties/cache/wii/cover/US/XXXXXX.png
/usr/share/rom-properties/cache/ds/coverS/US/XXXXXX.png

**Comment by bell07** [comment]
Created ebuild for the git/master version. With the latest version it works beter, thank you!
Seems the 1.6.1 is just broken for Linux.

3ds does work for me:
.cache/rom-properties/3ds/box/US/XXXX.jpg
.cache/rom-properties/3ds/cover/EN/XXXX.jpg
"box" contain some 0-byte files. For the same files the cover contains 0-byte file too.
The reason HTTP 404 Not Found.

wii - I did not setup the keys yet.

Gamecube: There is nothing in .cache/rom-properties/. But thunar ROM properties show me the banner for ISO file.

same for ds, nothing cached, but ROM properties show me the icons.

**Comment by bell07** [comment]
EDIT: My system is slow and requires some additional time ..

now the gamecube and ds files are downloaded in cache. 
.cache/rom-properties/ds/box/EN/XXXX.jpg
.cache/rom-properties/ds/cover/EN/XXXX.jpg

Gamecube + wii are both in wii folder
.cache/rom-properties/wii/cover/EN/XXXXXX.png
.cache/rom-properties/ds/disc/EN/XXXXXX.png

**Comment by GerbilSoft** [comment]
For the 404s on GameTDB, those files are missing. Someone needs to upload them.

There's a rather short timeout in rp-download to prevent the UI from freezing.

I'll probably make a v1.6.2 release soon, since there's a good number of low-level fixes.

**Comment by Amnesia1000** [comment]
So it looks like it's a Gnome exclusive issue. It does not take the thumbnails from .cache/rom-properties for me.

I will have to keep waiting to see if it is resolved

**Comment by GerbilSoft** [comment]
Nautilus runs all thumbnailers through the bubblewrap program, which restricts access to the user's home directory. Upstream rejected the patches to add .cache/rom-properties, so the /usr/share/cache option is the only one that'll work. I'll need to check it again to figure out why it wasn't working with my previous fixes.

**Comment by Amnesia1000** [comment]
> Nautilus runs all thumbnailers through the bubblewrap program, which restricts access to the user's home directory. Upstream rejected the patches to add .cache/rom-properties, so the /usr/share/cache option is the only one that'll work. I'll need to check it again to figure out why it wasn't working with my previous fixes.

Yes, I understood that, what I wanted to know is if the same thing happened to him once he updated. If it didn't work for him in the home folder, and he moved the files to the folder "/usr/share..." I wanted to know if it worked for him and how he had the folders organized to see if I had them correctly.

**Comment by dnmodder** [comment]
I get this error:
```
$ /usr/libexec/rp-download -v gba/title/P/BSTP7D.png
URL: https://rpdb.gerbilsoft.com/gba/title/P/BSTP7D.png
Cache Filename: /home/dnmodder/.cache/rom-properties/gba/title/P/BSTP7D.png
[1]    501863 invalid system call  /usr/libexec/rp-download -v gba/title/P/BSTP7D.png
```

**Comment by GerbilSoft** [comment]
This indicates that you have a newer glibc that's trying to use a new syscall that isn't whitelisted. (rp-download uses a seccomp() sandbox to only allow specific syscalls.)

What distribution are you using, and are you using rom-properties 1.7.3 or git master?

**Comment by dnmodder** [comment]
I'm using Manjaro, I get this error in both the git master version and v1.7.3, that's why 0 bytes thumbnails are created in the cache.

**Comment by GerbilSoft** [comment]
Can you rebuild it with debugging enabled (`-DCMAKE_BUILD_TYPE=Debug`) and see if it prints any "SYSCALL TRAP" messages?

**Comment by dnmodder** [comment]
Is this what you mean?
```
URL: https://rpdb.gerbilsoft.com/gba/title/P/BSTP7D.png
Cache Filename: /home/dnmodder/.cache/rom-properties/gba/title/P/BSTP7D.png
SYSCALL TRAP: [amd64] prctl()
SYSCALL TRAP: [amd64] mremap()
rp-download: malloc.c:3040: mremap_chunk: Assertion `aligned_OK (chunk2rawmem (p))' failed.
[1]    101157 abort (core dumped)  /usr/libexec/rp-download -v gba/title/P/BSTP7D.png
```

**Comment by GerbilSoft** [comment]
That is what I mean, and it seems that you're running into a code path that's using prctl() and mremap(). I'm not sure what would do that.

As a test, open src/rp-download/rp-download.cpp, go to line 397, and before the `-1` (end of whitelist), add the following line:

`SCMP_SYS(prctl), SCMP_SYS(mremap),`

Then recompile and test rp-download again.

**Comment by dnmodder** [comment]
It works, and in the file manager you can see the thumbnails of the roms, but what about that SYSCALL TRAP?
```
➜ /usr/libexec/rp-download -v gba/title/P/BSTP7D.png
URL: https://rpdb.gerbilsoft.com/gba/title/P/BSTP7D.png
Cache Filename: /home/dnmodder/.cache/rom-properties/gba/title/P/BSTP7D.png
SYSCALL TRAP: [amd64] ppoll()
Downloaded cache file for 'gba/title/P/BSTP7D.png': 13147 bytes.
```

**Comment by GerbilSoft** [comment]
Add `SCMP_SYS(ppoll),` to the same area.

I'm guessing the ppoll() failure isn't a serious error here.

**Comment by dnmodder** [comment]
![imagen](https://user-images.githubusercontent.com/12377383/115936465-4e75e980-a463-11eb-998a-4720947d8d01.png)
Okay, now can those changes be added in a commit or do I have to keep using them patch for a while?

**Comment by GerbilSoft** [comment]
I'm going to set up a Manjaro VM this weekend to figure out exactly why these syscalls are being used first. Once I figure it out, I'll add them to the syscall whitelist for rp-download, and possibly some other components.

**Comment by bell07** [comment]
Updated to  1.8.2 today. Now the issue seems to be solved.

**Comment by Clownacy** [comment]
I'm in the same boat: it seems to be working great now.

**Comment by GerbilSoft** [comment]
Closing as fixed in 1.8.
